### ISSUE TITLE
`alias`, `test_suite`, `genquery` should be implementable in pure Skylark

### ISSUE THREAD

**Issue by shs96c** [bazelbuild/bazel#11458]
### Description of the problem / feature request:

It should be possible to express every native rule of Bazel using Skylark, decoupling the underlying build engine from the implementation of rules.

### Feature requests: what underlying problem are you trying to solve with this feature?

The bazel team clearly believes that there should be a decoupling of the build engine and the rule sets. This approach is being followed for the language rules (eg `rules_java`, etc), but there are some intrinsic rules that are currently impossible to express using Starlark, and this causes problems (eg #11456 and #11455)

As an example, the simplest way of implementing `alias` would be to identify all the providers offered by a rule, and then re-exporting those as the output of the pure-starlark `alias` implementation. Doing so is hard because:

 1. There's no easy way to list all the providers of a rule
 2. All outputs of a rule must be created by that rule

However, `2` is hard to achieve because symlinking fails due to #11453 

### What's the output of `bazel info release`?

3.1.0

**Comment by sdtwigg** [comment]
I am unsure if genquery actually belongs on this list.

In theory (roughly, as I am just pondering this right now), a Starlark aspect should be able to traverse the given universe to collect a dependency graph. The Starlark genquery implementation would then consume that information as-needed. In practice, this is unlikely to be especially performant as the native implementation can very carefully skirt the line of hermiticity by hooking into the native query engine.

Alias and test_suite, especially currently, are extremely special rules.

The future of alias is itself debated (see #10893). My personal technical opinion is that alias itself is reasonable as a primitive (if the feature is even necessary long-term). Otherwise, a Starlark version would require allowing rules to read undeclared providers. I fear this would lead to maintenance headaches for Starlark rules in general if abused.

test_suite is special because it is specially expanded in the loading phase (along with other places that must be kept in sync) when deciding what the top-level targets to be expanded are. This is why features like #11455 would be challenging. A lot of top-level target expansion and test discovery would have to be re-written. A Starlarkified version would mean allowing Starlark code to hook in to that expansion.

**Comment by lberki** [comment]
I'll close this as "would be actively harmful if implemented".

In general, yes, all rules should be implementable in Starlark except "very special" ones, which the above rules are:

`genquery` needs to have access to the Skyframe evaluation machinery in order to be able to access the machinery that does `bazel query`. Giving access to this for everyone would be harmful.

`test_suite` is special-cased as discussed in (#10893). I'm not particularly fond of that situation but it will realistically stay that way for the foreseeable future. AFAIU @sdtwigg is working on making it a little less special, but things like its implicit dependency on every test rule in the same package will still be problems.

Doing `alias` in Starlark, but it would require relaxing some constraints that are there for a good reason, as @sdtwigg said. In addition, in order to provide nice error messages (like "`target //a:b (alias to //c:d) is not allowed here"), it's also special-cased in core code.

In summary, I think this would do more harm than good.
